### PR TITLE
[Updated #259] Add interface changes needed for gcc-trunk and testcases.

### PR DESCRIPTION
**Pull request by vehre**
Same description as in #259.
Sorry for the noise.

### CODE REVIEW

File: src/tests/unit/init_register/CMakeLists.txt [hunk 1]
```diff
@@ -4,14 +4,26 @@ target_link_libraries(initialize_mpi OpenCoarrays)
...
+target_link_libraries(register_alloc_comp_3 OpenCoarrays)
 
 add_executable(allocate_as_barrier allocate_as_barrier.f90)
 target_link_libraries(allocate_as_barrier OpenCoarrays)
 
 add_executable(allocate_as_barrier_proc allocate_as_barrier_proc.f90)
 target_link_libraries(allocate_as_barrier_proc OpenCoarrays)
+
+add_executable(check_remote_alloced_comp_1 check_remote_alloced_comp_1.f90)
+target_link_libraries(check_remote_alloced_comp_1 OpenCoarrays)
```
Comment (zbeekman) [inline src/tests/unit/init_register/CMakeLists.txt]: @vehre This is failing to compile with GCC 6.2. I'm assuming that this test is only valid for GCC trunk/7.x, correct? If so I'll wrap the CMake for conditional compilation when GCC >= 7.

File: src/tests/unit/init_register/CMakeLists.txt [hunk 2]
```diff
@@ -4,14 +4,26 @@ target_link_libraries(initialize_mpi OpenCoarrays)
...
 
-add_executable(register_rename_me register_rename_me.f90)
-target_link_libraries(register_rename_me OpenCoarrays)
+add_executable(register_vector register_vector.f90)
+target_link_libraries(register_vector OpenCoarrays)
 
-add_executable(register_rename_me_too register_rename_me_too.f90)
-target_link_libraries(register_rename_me_too OpenCoarrays)
+add_executable(register_alloc_vector register_alloc_vector.f90)
+target_link_libraries(register_alloc_vector OpenCoarrays)
```
Comment (zbeekman) [inline src/tests/unit/init_register/CMakeLists.txt]: Thanks so much for fixing these. It's been on my todo list for too long.

File: src/tests/unit/init_register/CMakeLists.txt [hunk 3]
```diff
@@ -4,14 +4,26 @@ target_link_libraries(initialize_mpi OpenCoarrays)
...
+
+add_executable(register_alloc_comp_1 register_alloc_comp_1.f90)
+target_link_libraries(register_alloc_comp_1 OpenCoarrays)
+
+add_executable(register_alloc_comp_2 register_alloc_comp_2.f90)
+target_link_libraries(register_alloc_comp_2 OpenCoarrays)
+
+add_executable(register_alloc_comp_3 register_alloc_comp_3.f90)
+target_link_libraries(register_alloc_comp_3 OpenCoarrays)
```
Comment (zbeekman) [inline src/tests/unit/init_register/CMakeLists.txt]: Do these tests require GCC trunk/7.x to compile? To run? If so I'll wrap them appropriately for conditional inclusion so long as GCC >= 7

**Comment by zbeekman** [comment]
If you confirm which tests require GCC trunk/7 to compile and run I can update the PR so that they are conditionally included/compiled when building with GCC 7

File: src/tests/unit/init_register/CMakeLists.txt (hunk 3 above)
Comment (vehre) [inline src/tests/unit/init_register/CMakeLists.txt]: Yes, of course. Sorry, I didn't think of that. Furthermore would I have to figure a way of how to it. Seems you already know, so please go ahead.

File: src/tests/unit/init_register/CMakeLists.txt (hunk 1 above)
Comment (vehre) [inline src/tests/unit/init_register/CMakeLists.txt]: Same as for above: Only for gcc-7.
And it will not run before #260 is done.

**Comment by vehre** [comment]
All of the new tests require gcc-7. The check_remote_alloced_comp_1 one requires even latest gcc-7.

**Comment by zbeekman** [comment]
OK, unless there are any objections, I'm going to cherry pick commits into a new branch and then add the tests that will run against GCC via Docker, and then add the guards against trying to execute the added tests when GCC 7 isn't available. I'll link to that here once I have done it. If I'm not mistaken @afanfa is expected to be able to take a look at this Friday or over the weekend, so I'll try to have things tidied and ready by then.

**Comment by vehre** [comment]
Well, you just be needed to add the checks for gcc v7 into the CMakeLists.txt for the new tests and should be it.

**Comment by zbeekman** [comment]
> Well, you just be needed to add the checks for gcc v7 into the CMakeLists.txt for the new tests and should be it.

Yes, but I would like to test OpenCoarrays against GCC trunk on our CI infrastructure anyway, so this will help verify that the tests run correctly... not strictly needed, but if we have time while waiting on @afanfa it won't hurt.

**Comment by afanfa** [comment]
I'm cloning Andre's repository. I want to take a closer look to the new register/deregister functions.

**Comment by afanfa** [comment]
@vehre , I've tried to test the asynchronous allocation of allocatable components but apparently the current gcc trunk doesn't support it. Is there a gcc branch with this support?

**Comment by vehre** [comment]
It is on trunk.

**Comment by vehre** [comment]
Can you be more precise what you expect to work and doesn't?

**Comment by afanfa** [comment]
@vehre , the program I tried hangs as it used to do in the past before the patch. I'm downloading a fresh new gcc-trunk on my computer at work. I'll let you know more precisely soon.

**Comment by vehre** [comment]
Did you do "your part" of the job already?
I have not made the necessary changes to opencoarray's caf_register to allow asynchronous allocation. I know to less about MPI-3 for that.

**Comment by zbeekman** [comment]
@fanfa: just FYI, I have a nightly build of GCC trunk available as a docker image: https://hub.docker.com/r/zbeekman/nightly-gcc-trunk-docker-image/ and it is only 176MB so it should be quick and easy to get up and running. I've been building OpenCoarrays in that image... first building MPICH with `./install.sh -p mpich -f gfortran -c gcc -C g++ -j 4` and then using the CMake script or Damian's installation script to compile OpenCoarrays. Even if you don't have docker installed yet, it may take less time to install docker and pull the image than rebuild GCC... 

Just a suggestion, feel free to take it or leave it.

**Comment by zbeekman** [comment]
I think @vehre is referring to some work that still needs to be done outlined here: #260

**Comment by afanfa** [comment]
OK, the compiler produces the right code. The pull request looks good to me.

**Comment by zbeekman** [comment]
@ afanfa, if you put `looks good to me`, `LGTM`, `:zap:`, `:shipit:` at the beginning of a line of your comment, you can :sparkles: *OFFICIALLY* :sparkles: sign off on this PR.

Like this:

LGTM

😄

**Comment by zbeekman** [comment]
@afanfa @rouson @vehre I'm going to make the changes to CMake and also cherry pick these relevant commits into a new branch and PR, to clean things up a little bit, unless anyone objects.

**Comment by afanfa** [comment]
LGTM

**Comment by vehre** [comment]
No objections from side. I just want to announce more testcases (at least one).

File: src/mpi/mpi_caf.c [hunk 4]
```diff
@@ -584,8 +585,6 @@ PREFIX (register) (size_t size, caf_register_t type, caf_token_t *token,
 
   if (type != CAF_REGTYPE_COARRAY_ALLOC_REGISTER_ONLY)
     {
-//      PREFIX(sync_all) (NULL,NULL,0);
-
```
Comment (zbeekman) [inline src/mpi/mpi_caf.c]: @vehre I'm a bit confused, the contents of this commit don't appear to match the description. Perhaps you forgot to `git add` some files before committing? (Also, if you didn't already know, adding th `-v` flag to `git commit` will show the diff/patch in your editor when composing the commit message...

Comment (zbeekman) [inline src/mpi/mpi_caf.c]: commit https://github.com/sourceryinstitute/opencoarrays/pull/261/commits/7e917b2d429306114f4c80b4ce723906bea27c8f to be precise

**Comment by zbeekman** [comment]
> I just want to announce more testcases (at least one).

@vehre: Where? If you have them you can just `git push <origin> vehre/async_comp_alloc` and they will be added to this pull request. Or if they are on another branch of your fork let me know which and I can pull them over to my cleaned up pull request.

File: src/mpi/mpi_caf.c (hunk 4 above)
Comment (zbeekman) [inline src/mpi/mpi_caf.c]: If there are additional changes that were not `git add`ed properly, just create a new commit with them and push your branch to github. This pull request will be automatically updated with those additional changes.

Comment (vehre) [inline src/mpi/mpi_caf.c]: No, that commit message and the changeset is absolutely correct. You don't see what is done in or by gfortran.

The comment I have added was to tell @afanfa what he already knows: He's got work there to do.

The additional testcases are not yet complete. Some of them are only in my mind or sleeping on my hard disc. I have to polish them before I can commit. But at the moment gfortran needs a lot of time.

**Comment by zbeekman** [comment]
Ah, I see, sorry for my confusion.... well I'll see if I can clarify the
commit message when I rebase to make this clearer. Just give me a heads up
when you have the additional tests and I can add them via a PR or email
patch or whatever.

**Comment by afanfa** [comment]
@vehre @zbeekman , if adding the tests takes more than one day could you please merge this pull request as soon as possible? I would like to start working on master.
Thanks.

**Comment by vehre** [comment]
@zbeekman go ahead with merging. Adding the test *will* take more than one day.

**Comment by zbeekman** [comment]
@afanfa #276 supersedes this PR: I've cleaned it up a little and wrapped the tests so they are only built and run with GCC >= 7. Please remember that you should work on a topic/feature branch and **not** *directly* on `master`. You don't even have to wait for #276 to hit master, you can just create a branch from `vehre/async-alloc-comp-support` which will be merged into master, and it will be as if you were working on master after the merge.

**Pull request closed without merging**